src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs

```diff
@@ -0,0 +1,386 @@
+        if !exists {
+            return Ok(Vec::new());
+        }
+
+        let q = r#"
+            SELECT timeline, GREATEST(read_ts, write_ts) as ts FROM timestamp_oracle;
+        "#;
+        let statement = client.prepare(q).await?;
+        let rows = client
+            .query(&statement, &[])
+            .await
+            .expect("todo: add internal retries");
+
+        let result = rows
+            .into_iter()
+            .map(|row| {
+                let timeline: String = row.try_get("timeline").expect("missing timeline column");
+                let ts: i64 = row.try_get("ts").expect("missing ts column");
+                let ts = Timestamp::try_from(ts).expect("we don't use negative timestamps");
+
+                (timeline, ts)
+            })
+            .collect::<Vec<_>>();
+
+        Ok(result)
+    }
+}
+
+// NOTE: This implementation is tied to [`mz_repr::Timestamp`]. We could change
+// that, and also make the types we store in the backing "Postgres" table
+// generic. But in practice we only use oracles for [`mz_repr::Timestamp`] so
+// don't do that extra work for now.
+#[async_trait(?Send)]
+impl<N> TimestampOracle<Timestamp> for PostgresTimestampOracle<N>
+where
+    N: GenericNowFn<Timestamp> + 'static,
+{
+    #[tracing::instrument(name = "oracle::write_ts", level = "debug", skip_all)]
+    async fn write_ts(&mut self) -> WriteTimestamp<Timestamp> {
+        let proposed_next_ts = self.next.now();
+        let proposed_next_ts =
+            i64::try_from(proposed_next_ts).expect("timestamp does not fit into i64");
+
+        let q = r#"
+            UPDATE timestamp_oracle SET write_ts = GREATEST(write_ts+1, $2)
+                WHERE timeline = $1
+            RETURNING write_ts;
+        "#;
+        let client = self
+            .get_connection()
+            .await
+            .expect("todo: add internal retries");
+        let statement = client
+            .prepare_cached(q)
+            .await
+            .expect("todo: add internal retries");
+        let result = client
+            .query_one(&statement, &[&self.timeline, &proposed_next_ts])
+            .await
+            .expect("todo: add internal retries");
+
+        let write_ts: i64 = result.try_get("write_ts").expect("missing column write_ts");
+        let write_ts = Timestamp::try_from(write_ts).expect("we don't use negative timestamps");
+
+        debug!(write_ts = ?write_ts, "write_ts");
+
+        let advance_to = write_ts.step_forward();
+
+        WriteTimestamp {
+            timestamp: write_ts,
+            advance_to,
+        }
+    }
+
+    #[tracing::instrument(name = "oracle::peek_write_ts", level = "debug", skip_all)]
+    async fn peek_write_ts(&self) -> Timestamp {
+        let q = r#"
+            SELECT write_ts FROM timestamp_oracle
+                WHERE timeline = $1;
+        "#;
+        let client = self
+            .get_connection()
+            .await
+            .expect("todo: add internal retries");
```

Comment:
This one too is being hit in the CI

```
platform-checks-materialized-1  | coordinator' panicked at 'todo: add internal retries: Backend(Error { kind: Connect, cause: Some(Custom { kind: Uncategorized, error: "failed to lookup address information: Name or service not known" }) })cluster-s2-replica-s2: 2023-10-10T10:29:35.251579Z  INFO mz_persist_client::internal::machine: external operation fetch_state::scan failed, retrying in 64ms: indeterminate: error connecting to server: failed to lookup address information: Name or service not known
platform-checks-materialized-1  | ', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0ad39b9e651b723c8-1/materialize/tests/src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs:318:14
```

In other words, I think we can rely on the existing tests once all the retry logic has been implemented.